README.md
nv-cipher
=========

Overview
--------

Cipher with encoder/decoder. BinaryEncoder and BinaryDecoder
of [Apache Commons Codec](http://commons.apache.org/proper/commons-codec/)
such as Base64, Hex and BinaryCodec can be used.

As a subclass of CodecCipher, AESCipher is contained which is dedicated to AES.


License
-------

Apache License, Version 2.0


Download
--------

    git clone https://github.com/TakahikoKawasaki/nv-cipher.git


JavaDoc
-------

[nv-cipher JavaDoc](http://TakahikoKawasaki.github.com/nv-cipher/)


Example
-------

```java
// Create a cipher with a secret key.
AESCipher cipher = new AESCipher().setKey("secret key", "initial vector");

// Encryption & decryption.
// 'plaintext' and 'decrypted' have the same value.
String plaintext = "plain text";
String encrypted = cipher.encrypt(plaintext);
String decrypted = cipher.decrypt(encrypted);

// In the above example, 'encrypted' is encoded by Base64 (default).
// If you want to change the format, use setCoder method.
// For example, to change the format to hexadecimal:
Hex hex = new org.apache.commons.codec.binary.Hex();
cipher.setCoder(hex);

// Binary representation (only "0"s and "1"s) also can be used.
BinaryCodec binary = new org.apache.commons.codec.BinaryCodec();
cipher.setCoder(binary);

// Coder can be specified as a constructor parameter.
cipher = new AESCipher(hex);

// If you want, an encoder and a decoder can be set separately.
cipher.setEncoder(hex);
cipher.setDecoder(hex);
```


Maven
-----

```xml
<dependency>
    <groupId>com.neovisionaries</groupId>
    <artifactId>nv-cipher</artifactId>
    <version>1.4</version>
</dependency>
```


See Also
--------

* [javax.crypto.Cipher](http://docs.oracle.com/javase/7/docs/api/javax/crypto/Cipher.html)
* [Apache Commons Codec](http://commons.apache.org/proper/commons-codec/)


Author
------

Takahiko Kawasaki @ Authlete, Inc.



CHANGES.md
CHANGES
=======

1.4 (2020-12-11)
----------------

- `AESCipher` class
    * Added `DEFAULT_KEY_SIZE` class variable.
    * Added `DEFAULT_TRANSFORMATION` class variable.
    * Added `setKey(byte[] key, byte[] iv, int keySize)` method.
    * Added `setKey(String key, byte[] iv, int keySize)` method.
    * Added `setKey(String key, String iv, int keySize)` method.
    * Added `setKey(String key, int keySize)` method.
    * Added `setKey(byte[] key, String iv, int keySize)` method.
    * Added `setKey(byte[] key, int keySize)` method.

- `pom.xml`
    * Changed `source` and `target` versions of Java from 1.5 to 8.


src/test/java/com/neovisionaries/security/AESCipherTest.java
/*
 * Copyright (C) 2014-2020 Neo Visionaries Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.neovisionaries.security;


import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import org.apache.commons.codec.BinaryDecoder;
import org.apache.commons.codec.BinaryEncoder;
import org.apache.commons.codec.binary.BinaryCodec;
import org.apache.commons.codec.binary.Hex;
import org.apache.commons.codec.net.QuotedPrintableCodec;
import org.apache.commons.codec.net.URLCodec;
import org.junit.Ignore;
import org.junit.Test;


/**
 * Tests for {@link AESCipher}.
 *
 * @author Takahiko Kawasaki
 */
public class AESCipherTest
{
    private static final boolean DEBUG = false;


    private <TCoder extends BinaryEncoder & BinaryDecoder> void doTest(String plain, String key, String iv, TCoder coder)
    {
        doTest(plain, key, iv, AESCipher.DEFAULT_KEY_SIZE, coder);
    }


    private <TCoder extends BinaryEncoder & BinaryDecoder> void doTest(String plain, String key, String iv, int keySize, TCoder coder)
    {
        AESCipher cipher = new AESCipher();

        if (coder != null)
        {
            cipher.setCoder(coder);
        }

        cipher.setKey(key, iv, keySize);

        // Encrypt.
        String encrypted = cipher.encrypt(plain);

        if (iv == null)
        {
            byte[] iv2 = cipher.getCipher().getIV();
            cipher.setKey(key, iv2);
        }

        // Decrypt.
        String decrypted = cipher.decrypt(encrypted);

        if (DEBUG)
        {
            System.out.println("----------");
            System.out.println("codec     = " + ((coder != null) ? coder.getClass().getSimpleName() : "default"));
            System.out.println("plain     = " + plain);
            System.out.println("encrypted = " + encrypted);
            System.out.println("decrypted = " + decrypted);
        }

        assertNotEquals(plain, encrypted);
        assertEquals(plain, decrypted);
    }


    @Test
    public void test1()
    {
        doTest("hello", "key", "iv", null);
    }


    @Test
    public void test2()
    {
        doTest("hello", "1234567890123456", "1234567890123456", new Hex());
    }


    @Test
    public void test3()
    {
        doTest("hello", "12345678901234567890", "12345678901234567890", new BinaryCodec());
    }


    @Test
    public void test4()
    {
        doTest("hello", "key", null, new URLCodec());
    }


    @Test
    public void test5()
    {
        doTest("hello", "1234567890123456", null, new QuotedPrintableCodec());
    }


    @Test
    public void test6()
    {
        doTest("hello", "12345678901234567890", null, null);
    }


    @Test
    public void test7()
    {
        doTest("hello", null, null, null);
    }


    @Ignore("This fails if JVM does not support 256-bit key size for AES.")
    @Test
    public void test8()
    {
        doTest("hello", "0123456789ABCDEF0123456789ABCDEF", "0123456789ABCDEF", 32, null);
    }
}


src/main/java/com/neovisionaries/security/Utils.java
/*
 * Copyright (C) 2014 Neo Visionaries Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.neovisionaries.security;


import java.io.UnsupportedEncodingException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;


/**
 * Utility methods for internal implementation.
 *
 * @author Takahiko Kawasaki
 */
class Utils
{
    private Utils()
    {
    }


    /**
     * Get bytes by {@code input.getBytes("UTF-8")}.
     */
    public static byte[] getBytesUTF8(String string)
    {
        if (string == null)
        {
            return null;
        }

        try
        {
            // Convert the string to a byte array encoded in UTF-8.
            return string.getBytes("UTF-8");
        }
        catch (UnsupportedEncodingException e)
        {
            // This won't happen.
            return null;
        }
    }


    /**
     * Build a {@code String} instance by {@code new String(input, "UTF-8")}.
     */
    public static String toStringUTF8(byte[] input)
    {
        if (input == null)
        {
            return null;
        }

        try
        {
            return new String(input, "UTF-8");
        }
        catch (UnsupportedEncodingException e)
        {
            // This won't happen.
            return null;
        }
    }


    public static byte[] ensureSize(byte[] data, int size)
    {
        if (data == null)
        {
            return new byte[size];
        }

        if (size <= data.length)
        {
            return data;
        }

        byte[] data2 = new byte[size];

        System.arraycopy(data, 0, data2, 0, data.length);

        return data2;
    }


    public static SecretKeySpec createSecretKeySpec(String key, String algorithm, int size)
    {
        return createSecretKeySpec(getBytesUTF8(key), algorithm, size);
    }


    public static SecretKeySpec createSecretKeySpec(byte[] key, String algorithm, int size)
    {
        key = ensureSize(key, size);

        return new SecretKeySpec(key, 0, size, algorithm);
    }


    public static IvParameterSpec createIvParameterSpec(String iv, int size)
    {
        return createIvParameterSpec(getBytesUTF8(iv), size);
    }


    public static IvParameterSpec createIvParameterSpec(byte[] iv, int size)
    {
        iv = ensureSize(iv, size);

        return new IvParameterSpec(iv, 0, size);
    }
}


src/main/java/com/neovisionaries/security/StandardCipherTransformations.java
/*
 * Copyright (C) 2014 Neo Visionaries Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.neovisionaries.security;


/**
 * List of standard cipher transformations which are valid as
 * arguments for {@link javax.crypto.Cipher#getInstance(String)}.
 *
 * <p>
 * The list here is a copy from Java SE 7 JavaDoc. Therefore,
 * they may not be supported in older Java SE environments.
 * </p>
 *
 * @see <a href="http://docs.oracle.com/javase/7/docs/api/javax/crypto/Cipher.html"
 *      >javax.crypto.Cipher</a>
 *
 * @author Takahiko Kawasaki
 */
public class StandardCipherTransformations
{
    /**
     * AES/CBC/NoPadding (128)
     */
    public static final String AES_CBC_NOPADDING = "AES/CBC/NoPadding";


    /**
     * AES/CBC/PKCS5Padding (128)
     */
    public static final String AES_CBC_PKCS5PADDING = "AES/CBC/PKCS5Padding";


    /**
     * AES/ECB/NoPadding (128)
     */
    public static final String AES_ECB_NOPADDING = "AES/ECB/NoPadding";


    /**
     * AES/ECB/PKCS5Padding (128)
     */
    public static final String AES_ECB_PKCS5PADDING = "AES/ECB/PKCS5Padding";


    /**
     * DES/CBC/NoPadding (56)
     */
    public static final String DES_CBC_NOPADDING = "DES/CBC/NoPadding";


    /**
     * DES/CBC/PKCS5Padding (56)
     */
    public static final String DES_CBC_PKCS5PADDING = "DES/CBC/PKCS5Padding";


    /**
     * DES/ECB/NoPadding (56)
     */
    public static final String DES_ECB_NOPADDING = "DES/ECB/NoPadding";


    /**
     * DES/ECB/PKCS5Padding (56)
     */
    public static final String DES_ECB_PKCS5PADDING = "DES/ECB/PKCS5Padding";


    /**
     * DESede/CBC/NoPadding (168)
     */
    public static final String DESEDE_CBC_NOPADDING = "DESede/CBC/NoPadding";


    /**
     * DESede/CBC/PKCS5Padding (168)
     */
    public static final String DESEDE_CBC_PKCS5PADDING = "DESede/CBC/PKCS5Padding";


    /**
     * DESede/ECB/NoPadding (168)
     */
    public static final String DESEDE_ECB_NOPADDING = "DESede/ECB/NoPadding";


    /**
     * DESede/ECB/PKCS5Padding (168)
     */
    public static final String DESEDE_ECB_PKCS5PADDING = "DESede/ECB/PKCS5Padding";


    /**
     * RSA/ECB/PKCS1Padding (1024, 2048)
     */
    public static final String RSA_ECB_PKCS1PADDING = "RSA/ECB/PKCS1Padding";


    /**
     * RSA/ECB/OAEPWithSHA-1AndMGF1Padding (1024, 2048)
     */
    public static final String RSA_ECB_OAEPWITHSHA1ANDMGF1PADDING = "RSA/ECB/OAEPWithSHA-1AndMGF1Padding";


    /**
     * RSA/ECB/OAEPWithSHA-256AndMGF1Padding (1024, 2048)
     */
    public static final String RSA_ECB_OAEPWITHSHA256ANDMGF1PADDING = "RSA/ECB/OAEPWithSHA-256AndMGF1Padding";


    private StandardCipherTransformations()
    {
    }
}


src/main/java/com/neovisionaries/security/AESCipher.java
/*
 * Copyright (C) 2014-2020 Neo Visionaries Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.neovisionaries.security;


import static com.neovisionaries.security.StandardCipherTransformations.AES_CBC_PKCS5PADDING;
import javax.crypto.SecretKey;
import javax.crypto.spec.IvParameterSpec;
import org.apache.commons.codec.BinaryDecoder;
import org.apache.commons.codec.BinaryEncoder;
import org.apache.commons.codec.binary.Base64;


/**
 * Cipher using {@code "AES/CBC/PKCS5Padding"}.
 *
 * <pre style="background-color: #EEEEEE; margin: 2em; border: 1px solid black; padding: 0.5em;">
 * <span style="color: darkgreen;">// Create a cipher with a secret key.</span>
 * AESCipher cipher = new {@link #AESCipher()}.{@link #setKey(String, String) setKey}(<span style="color: darkred;">"secret key"</span>, <span style="color: darkred;">"initial vector"</span>);
 *
 * <span style="color: darkgreen;">// Encryption &amp; decryption.
 * // 'plaintext' and 'decrypted' have the same value.</span>
 * String plaintext = <span style="color: darkred;">"plain text"</span>;
 * String encrypted = cipher.{@link #encrypt(String) encrypt(plaintext)};
 * String decrypted = cipher.{@link #decrypt(String) decrypt(encrypted)};
 *
 * <span style="color: darkgreen;">// In the above example, 'encrypted' is encoded by Base64 (default).
 * // If you want to change the format, use {@code setCoder} method.
 * // For example, to change the format to hexadecimal:</span>
 * Hex hex = new org.apache.commons.codec.binary.Hex();
 * cipher.setCoder(hex);
 *
 * <span style="color: darkgreen;">// Binary representation (only "0"s and "1"s) also can be used.</span>
 * BinaryCodec binary = new org.apache.commons.codec.BinaryCodec();
 * cipher.setCoder(binary);
 *
 * <span style="color: darkgreen;">// Coder can be specified as a constructor parameter.</span>
 * cipher = new AESCipher(hex);
 *
 * <span style="color: darkgreen;">// If you want, an encoder and a decoder can be set separately.</span>
 * cipher.{@link #setEncoder(BinaryEncoder) setEncoder(hex)};
 * cipher.{@link #setDecoder(BinaryDecoder) setDecoder(hex)};
 * </pre>
 *
 * <pre style="background-color: #EEEEEE; margin: 2em; border: 1px solid black; padding: 0.5em;">
 * <span style="color: darkgreen;">// Another example which performs encryption without initial vector.</span>
 * String secretkey = <span style="color: darkred;">"secret key"</span>;
 * String plaintext = <span style="color: darkred;">"plain text"</span>;
 *
 * <span style="color: darkgreen;">// Create and set up without initial vector.</span>
 * AESCipher cipher = new AESCipher().setKey(secretkey);
 *
 * <span style="color: darkgreen;">// Encrypt.</span>
 * String encrypted = cipher.encrypt(plaintext);
 *
 * <span style="color: darkgreen;">// Get the auto-generated initial vector.</span>
 * byte[] iv = cipher.getCipher().getIV();
 *
 * <span style="color: darkgreen;">// Decryption requires initial vector.</span>
 * cipher.setKey(secretkey, iv);
 *
 * <span style="color: darkgreen;">// Decrypt.</span>
 * String decrypted = cipher.decrypt(encrypted);
 * </pre>
 *
 * @author Takahiko Kawasaki
 */
public class AESCipher extends CodecCipher
{
    /**
     * The default key size. The value is 16, meaning 128 bits.
     *
     * @since 1.4
     */
    public static final int DEFAULT_KEY_SIZE = 16;


    /**
     * The default transformation. The value is "AES/CBC/PKCS5Padding".
     *
     * @since 1.4
     */
    public static final String DEFAULT_TRANSFORMATION = AES_CBC_PKCS5PADDING;


    /**
     * The initial vector. The size of initial vectors for AES is always 16.
     *
     * @since 1.4
     */
    private static final int INITIAL_VECTOR_SIZE = 16;


    /**
     * Constructor.
     *
     * <p>
     * This constructor just performs {@link CodecCipher#CodecCipher(String)
     * super("AES/CBC/PKCS5Padding")}.
     * </p>
     */
    public AESCipher()
    {
        super(DEFAULT_TRANSFORMATION);
    }


    /**
     * Constructor.
     *
     * <p>
     * This constructor just performs {@link CodecCipher#CodecCipher(String)
     * super(transformation)}.
     * </p>
     *
     * @since 1.3
     */
    public AESCipher(String transformation)
    {
        super(transformation);
    }


    /**
     * Constructor with an encoder and a decoder.
     *
     * <p>
     * This constructor just performs {@link CodecCipher#CodecCipher(String,
     * BinaryEncoder, BinaryDecoder) super("AES/CBC/PKCS5Padding", encoder, decoder)}.
     * </p>
     *
     * @param encoder
     *         An encoder used in {@link #encrypt(String) encrypt(String)} and
     *         {@link #encrypt(byte[]) encrypt(byte[])} to encode an encrypted byte array.
     *         If {@code null} is given, {@link Base64} is used as the default
     *         encoder.
     *
     * @param decoder
     *         A decoder used in {@link #decrypt(String) decrypt(String)} and
     *         {@link #decrypt(byte[]) decrypt(byte[])} to decode an encoded input byte array.
     *         If {@code null} is given, {@link Base64} is used as the default
     *         decoder.
     */
    public AESCipher(BinaryEncoder encoder, BinaryDecoder decoder)
    {
        super(DEFAULT_TRANSFORMATION, encoder, decoder);
    }


    /**
     * Constructor with a transformation, an encoder and a decoder.
     *
     * <p>
     * This constructor just performs {@link CodecCipher#CodecCipher(String,
     * BinaryEncoder, BinaryDecoder) super(transformation, encoder, decoder)}.
     * </p>
     *
     * @param encoder
     *         An encoder used in {@link #encrypt(String) encrypt(String)} and
     *         {@link #encrypt(byte[]) encrypt(byte[])} to encode an encrypted byte array.
     *         If {@code null} is given, {@link Base64} is used as the default
     *         encoder.
     *
     * @param decoder
     *         A decoder used in {@link #decrypt(String) decrypt(String)} and
     *         {@link #decrypt(byte[]) decrypt(byte[])} to decode an encoded input byte array.
     *         If {@code null} is given, {@link Base64} is used as the default
     *         decoder.
     *
     * @since 1.3
     */
    public AESCipher(String transformation, BinaryEncoder encoder, BinaryDecoder decoder)
    {
        super(transformation, encoder, decoder);
    }


    /**
     * Constructor with a coder.
     *
     * <p>
     * This constructor just performs {@code super("AES/CBC/PKCS5Padding", coder)}.
     * </p>
     *
     * @param coder
     *         A coder which works as both an encoder and a decoder.
     *         If {@code null} is given, {@link Base64} is used as the
     *         default coder.
     */
    public <TCoder extends BinaryEncoder & BinaryDecoder> AESCipher(TCoder coder)
    {
        super(DEFAULT_TRANSFORMATION, coder);
    }


    /**
     * Constructor with a transformation and a coder.
     *
     * <p>
     * This constructor just performs {@code super(transformation, coder)}.
     * </p>
     *
     * @param coder
     *         A coder which works as both an encoder and a decoder.
     *         If {@code null} is given, {@link Base64} is used as the
     *         default coder.
     *
     * @since 1.3
     */
    public <TCoder extends BinaryEncoder & BinaryDecoder> AESCipher(String transformation, TCoder coder)
    {
        super(transformation, coder);
    }


    /**
     * Set cipher initialization parameters.
     *
     * <p>
     * This method is an alias of {@link #setInit(java.security.Key,
     * java.security.spec.AlgorithmParameterSpec) setInit(key, iv)}.
     * </p>
     *
     * @param key
     *         Secret key.
     *
     * @param iv
     *         Initial vector.
     *
     * @return
     *         {@code this} object.
     *
     * @throws IllegalArgumentException
     *         {@code key} is {@code null}.
     */
    public AESCipher setKey(SecretKey key, IvParameterSpec iv)
    {
        return (AESCipher)setInit(key, iv);
    }


    /**
     * Set cipher initialization parameters.
     *
     * <p>
     * This method is an alias of {@link #setKey(byte[], byte[], int)
     * setKey}<code>(key, iv, {@link #DEFAULT_KEY_SIZE})</code>.
     * </p>
     *
     * @param key
     *         Secret key.
     *
     * @param iv
     *         Initial vector.
     *
     * @return
     *         {@code this} object.
     */
    public AESCipher setKey(byte[] key, byte[] iv)
    {
        return setKey(key, iv, DEFAULT_KEY_SIZE);
    }


    /**
     * Set cipher initialization parameters. Other {@code setKey}
     * method variants except {@link #setKey(SecretKey, IvParameterSpec)}
     * eventually call this method.
     *
     * <p>
     * This method constructs a {@link SecretKey} instance and an
     * {@link IvParameterSpec} instance from the arguments,
     * and then calls {@link #setKey(SecretKey, IvParameterSpec)}.
     * </p>
     *
     * @param key
     *         Secret key. If {@code null} is given, {@code new byte[16]}
     *         is used. If not {@code null} and the length is less than 16,
     *         a byte array of size 16 is allocated and the content of
     *         {@code key} is copied to the newly allocated byte array,
     *         and the resultant byte array is used.
     *
     * @param iv
     *         Initial vector. If {@code null} is given, {@code null}
     *         is used, meaning that {@code IvParameterSepc} argument
     *         passed to {@link #setKey(SecretKey, IvParameterSpec)} is
     *         {@code null}. In that case, you will want to obtain the
     *         auto-generated initial vector by calling {@link #getCipher()
     *         getCipher()}{@code .}{@link javax.crypto.Cipher#getIV()
     *         getIV()} in order to decrypt the encrypted data.
     *
     *         <p>
     *         If {@code iv} is not {@code null} and the length is less
     *         than 16, a byte array of size 16 is allocated and the content
     *         of {@code iv} is copied to the newly allocated byte array,
     *         and the resultant byte array is used. Even if the length is
     *         greater than 16, only the first 16 bytes are used to construct
     *         an {@code IvParameterSpec} instance.
     *         </p>
     *
     * @param keySize
     *         The size of the secret key in bytes.
     *
     * @return
     *         {@code this} object.
     *
     * @since 1.4
     */
    public AESCipher setKey(byte[] key, byte[] iv, int keySize)
    {
        SecretKey secretKey  = Utils.createSecretKeySpec(key, getAlgorithm(), keySize);
        IvParameterSpec spec = null;

        if (iv != null)
        {
            spec = Utils.createIvParameterSpec(iv, INITIAL_VECTOR_SIZE);
        }

        return setKey(secretKey, spec);
    }


    /**
     * Set cipher initialization parameters.
     *
     * <p>
     * This method is an alias of {@link #setKey(String, byte[], int)
     * setKey}<code>(key, iv, {@link #DEFAULT_KEY_SIZE})</code>.
     * </p>
     *
     * @param key
     *         Secret key.
     *
     * @param iv
     *         Initial vector.
     *
     * @return
     *         {@code this} object.
     *
     * @since 1.2
     */
    public AESCipher setKey(String key, byte[] iv)
    {
        return setKey(key, iv, DEFAULT_KEY_SIZE);
    }


    /**
     * Set cipher initialization parameters.
     *
     * <p>
     * This method is an alias of {@link #setKey(byte[], byte[], int)}.
     * </p>
     *
     * @param key
     *         Secret key. The value is converted to a byte array
     *         by {@code key.getBytes("UTF-8")} and used as the
     *         first argument of {@link #setKey(byte[], byte[], int)}.
     *
     * @param iv
     *         Initial vector.
     *
     * @param keySize
     *         The size of the secret key in bytes.
     *
     * @return
     *         {@code this} object.
     *
     * @since 1.4
     */
    public AESCipher setKey(String key, byte[] iv, int keySize)
    {
        byte[] key2 = Utils.getBytesUTF8(key);

        return setKey(key2, iv, keySize);
    }


    /**
     * Set cipher initialization parameters.
     *
     * <p>
     * This method is an alias of {@link #setKey(String, String, int)
     * setKey}<code>(key, iv, {@link #DEFAULT_KEY_SIZE})</code>.
     * </p>
     *
     * @param key
     *         Secret key.
     *
     * @param iv
     *         Initial vector.
     *
     * @return
     *         {@code this} object.
     */
    public AESCipher setKey(String key, String iv)
    {
        return setKey(key, iv, DEFAULT_KEY_SIZE);
    }


    /**
     * Set cipher initialization parameters.
     *
     * <p>
     * This method is an alias of {@link #setKey(byte[], byte[], int)}.
     * </p>
     *
     * @param key
     *         Secret key. The value is converted to a byte array
     *         by {@code key.getBytes("UTF-8")} and used as the
     *         first argument of {@link #setKey(byte[], byte[], int)}.
     *
     * @param iv
     *         Initial vector. The value is converted to a byte array
     *         by {@code iv.getBytes("UTF-8")} and used as the
     *         second argument of {@link #setKey(byte[], byte[], int)}.
     *
     * @param keySize
     *         The size of the secret key in bytes.
     *
     * @return
     *         {@code this} object.
     *
     * @since 1.4
     */
    public AESCipher setKey(String key, String iv, int keySize)
    {
        byte[] key2 = Utils.getBytesUTF8(key);
        byte[] iv2  = Utils.getBytesUTF8(iv);

        return setKey(key2, iv2, keySize);
    }


    /**
     * Set cipher initialization parameters.
     *
     * <p>
     * This method is an alias of {@link #setKey(String, int)
     * setKey}<code>(key, {@link #DEFAULT_KEY_SIZE})</code>
     * </p>
     *
     * @param key
     *         Secret key.
     *
     * @return
     *         {@code this} object.
     */
    public AESCipher setKey(String key)
    {
        return setKey(key, DEFAULT_KEY_SIZE);
    }


    /**
     * Set cipher initialization parameters.
     *
     * <p>
     * This method is an alias of {@link #setKey(byte[], byte[], int)}.
     * </p>
     *
     * @param key
     *         Secret key. The value is converted to a byte array
     *         by {@code key.getBytes("UTF-8")} and used as the
     *         first argument of {@link #setKey(byte[], byte[], int)}.
     *
     * @param keySize
     *         The size of the secret key in bytes.
     *
     * @return
     *         {@code this} object.
     *
     * @since 1.4
     */
    public AESCipher setKey(String key, int keySize)
    {
        byte[] key2 = Utils.getBytesUTF8(key);

        return setKey(key2, (byte[])null, keySize);
    }


    /**
     * Set cipher initialization parameters.
     *
     * <p>
     * This method is an alias of {@link #setKey(byte[], String, int)
     * setKey}<code>(key, iv, {@link #DEFAULT_KEY_SIZE})</code>.
     * </p>
     *
     * @param key
     *         Secret key.
     *
     * @param iv
     *         Initial vector.
     *
     * @return
     *         {@code this} object.
     *
     * @since 1.2
     */
    public AESCipher setKey(byte[] key, String iv)
    {
        return setKey(key, iv, DEFAULT_KEY_SIZE);
    }


    /**
     * Set cipher initialization parameters.
     *
     * <p>
     * This method is an alias of {@link #setKey(byte[], byte[], int)}.
     * </p>
     *
     * @param key
     *         Secret key.
     *
     * @param iv
     *         Initial vector. The value is converted to a byte array
     *         by {@code iv.getBytes("UTF-8")} and used as the
     *         second argument of {@link #setKey(byte[], byte[], int)}.
     *
     * @param keySize
     *         The size of the secret key in bytes.
     *
     * @return
     *         {@code this} object.
     *
     * @since 1.4
     */
    public AESCipher setKey(byte[] key, String iv, int keySize)
    {
        byte[] iv2 = Utils.getBytesUTF8(iv);

        return setKey(key, iv2, keySize);
    }


    /**
     * Set cipher initialization parameters.
     *
     * <p>
     * This method is an alias of {@link #setKey(byte[], int)
     * setKey}<code>(key, {@link #DEFAULT_KEY_SIZE})</code>.
     * </p>
     *
     * @param key
     *         Secret key.
     *
     * @return
     *         {@code this} object.
     *
     * @since 1.2
     */
    public AESCipher setKey(byte[] key)
    {
        return setKey(key, DEFAULT_KEY_SIZE);
    }


    /**
     * Set cipher initialization parameters.
     *
     * <p>
     * This method is an alias of {@link #setKey(byte[], byte[], int)
     * setKey}<code>(key, (byte[])null, keySize)</code>.
     * </p>
     *
     * @param key
     *         Secret key.
     *
     * @param keySize
     *         The size of the secret key in bytes.
     *
     * @return
     *         {@code this} object.
     *
     * @since 1.4
     */
    public AESCipher setKey(byte[] key, int keySize)
    {
        return setKey(key, (byte[])null, keySize);
    }
}
